BootloaderCommonPkg/Library/LiteFvLib/FvLib.c

```diff
@@ -161,3 +275,70 @@ GetSectionByType (
   *SectionData = NULL;
   return EFI_NOT_FOUND;
 }
+
+
+/**
+  Load FV to preferred location and return entry point of SEC core.
+
+  This function will search whole FV to find SEC core file, and then check SEC core
+  file if relocation is required. If relocation is required, this function will copy
+  the whole FV to preferred location and SEC core entry point will be returned if
+  success.
+
+  @param[in]  FvBase                   Point to the boot firmware volume.
+  @param[in]  FvLength                 The actural length of FV.
+  @param[out] EntryPoint               The pointer to receive SecCore entry point.
+
+  @retval RETURN_SUCCESS               The FV is loaded successfully.
+  @retval Others                       Failed to load the FV.
+**/
+EFI_STATUS
+LoadFvImage (
+  IN  UINT32                            *FvBase,
+  IN  UINT32                             FvLength,
+  OUT VOID                             **EntryPoint
+  )
+{
+  EFI_STATUS                            Status;
+  EFI_FIRMWARE_VOLUME_HEADER           *FvHeader;
+  EFI_FFS_FILE_HEADER                  *SecCoreFile;
+  EFI_COMMON_SECTION_HEADER            *Section;
+  UINT32                                PreferredBase;
+  UINT32                                SecCoreImageBase;
+  INT32                                 Gap;
+
+  FvHeader = (EFI_FIRMWARE_VOLUME_HEADER *)FvBase;
+
+  // Find SecCore file first
+  Status = GetFfsFileByType (FvHeader, EFI_FV_FILETYPE_SECURITY_CORE, 0, &SecCoreFile);
+  if (EFI_ERROR (Status)) {
+    return EFI_NOT_FOUND;
+  }
+
+  // Find PE32 or TE section in SecCore file
+  Status = GetSectionByType (SecCoreFile, EFI_SECTION_PE32, 0, &Section);
+  if (Status == EFI_NOT_FOUND) {
+    Status = GetSectionByType (SecCoreFile, EFI_SECTION_TE, 0, &Section);
+  }
+  if (EFI_ERROR (Status)) {
+    return EFI_NOT_FOUND;
+  }
+
+  // Check preferred image base
+  SecCoreImageBase = (UINTN)Section;
+  Status = PeCoffGetPreferredBase ((VOID *)SecCoreImageBase, &PreferredBase);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  // Move FV to preferred base if required
+  Gap = (UINT32)(UINTN)PreferredBase - SecCoreImageBase;
+  if (Gap != 0) {
+    CopyMem ((UINT8 *)FvHeader + Gap, FvHeader, FvLength);
```

Comment:
We should probably check for destination/source partial overlap in this case (or use something which does so like a `memmove`)